apiserver/common/upgradeseries.go

```diff
@@ -0,0 +1,224 @@
+			continue
+		}
+		watcherId = u.resources.Register(w)
+		result.Results[i].NotifyWatcherId = watcherId
+	}
+	return result, nil
+}
+
+// UpgradeSeriesStatus returns the current state of series upgrading
+// unit. If no upgrade is in progress an error is returned instead.
+func (u *UpgradeSeriesAPI) UpgradeSeriesStatus(args params.Entities) (params.UpgradeSeriesStatusResults, error) {
+	u.logger.Tracef("Starting UpgradeSeriesStatus with %+v", args)
+	result := params.UpgradeSeriesStatusResults{
+		Results: make([]params.UpgradeSeriesStatusResult, len(args.Entities)),
+	}
+	canAccess, err := u.accessUnitOrMachine()
+	if err != nil {
+		return params.UpgradeSeriesStatusResults{}, err
+	}
+	for i, entity := range args.Entities {
+		tag, err := names.ParseUnitTag(entity.Tag)
+		if err != nil {
+			result.Results[i].Error = ServerError(ErrPerm)
+			continue
+		}
+
+		if !canAccess(tag) {
+			result.Results[i].Error = ServerError(ErrPerm)
+			continue
+		}
+		unit, err := u.getUnit(tag)
+		if err != nil {
+			result.Results[i].Error = ServerError(err)
+			continue
+		}
+		status, err := unit.UpgradeSeriesStatus()
+		if err != nil {
+			result.Results[i].Error = ServerError(err)
+			continue
+		}
+		result.Results[i].Status = string(status)
+	}
+
+	return result, nil
+}
+
+func (u *UpgradeSeriesAPI) SetUpgradeSeriesStatus(args params.SetUpgradeSeriesStatusParams) (params.ErrorResults, error) {
+	u.logger.Tracef("Starting SetUpgradeSeriesStatus with %+v", args)
+	result := params.ErrorResults{
+		Results: make([]params.ErrorResult, len(args.Params)),
+	}
+	canAccess, err := u.accessUnit()
+	if err != nil {
+		return params.ErrorResults{}, err
+	}
+	for i, p := range args.Params {
+		//TODO[externalreality] refactor all of this, its being copied often.
+		tag, err := names.ParseUnitTag(p.Entity.Tag)
+		if err != nil {
+			result.Results[i].Error = ServerError(ErrPerm)
+			continue
+		}
+		if !canAccess(tag) {
+			result.Results[i].Error = ServerError(ErrPerm)
+			continue
+		}
+		unit, err := u.getUnit(tag)
+		if err != nil {
+			result.Results[i].Error = ServerError(err)
+			continue
+		}
+		status, err := model.ValidateUnitSeriesUpgradeStatus(p.Status)
+		if err != nil {
+			result.Results[i].Error = ServerError(err)
+			continue
+		}
+		err = unit.SetUpgradeSeriesStatus(status)
+		if err != nil {
+			result.Results[i].Error = ServerError(err)
+			continue
+		}
+	}
+
+	return result, nil
+}
+
+func (u *UpgradeSeriesAPI) getMachine(tag names.Tag) (UpgradeSeriesMachine, error) {
+	var id string
+	switch tag.Kind() {
+	case names.MachineTagKind:
+		id = tag.Id()
+	case names.UnitTagKind:
+		unit, err := u.backend.Unit(tag.Id())
+		if err != nil {
+
+		}
+		id, err = unit.AssignedMachineId()
+		if err != nil {
+			return nil, err
+		}
+	default:
+	}
+	return u.backend.Machine(id)
+}
+
+func (u *UpgradeSeriesAPI) getUnit(tag names.Tag) (UpgradeSeriesUnit, error) {
+	return u.backend.Unit(tag.Id())
+}
+
+func NewExternalUpgradeSeriesAPI(
```

Comment:
Do we need to comment this exported function?

Reply:
It's used by the apiserver/.../uniter to initialize this piece.  NewUpgradeSeriesAPI() will be used by the UpgradeSeriesAPI for the new agent.  Having the interfaces facilities testing.

Reply:
misread the original comment.